### PR TITLE
Update example packages to use new kpt

### PR DESCRIPTION
**Pull request by mengqiy**
This PR migrates all curated functions except apply-setters and search-replace.
I will do the migration for the contrib functions in a separate PR.
I added an imperative example for each function if possible.

There are a few more examples that are planed to be added https://github.com/GoogleContainerTools/kpt/issues/1989 in a separate PR.

fixes: https://github.com/GoogleContainerTools/kpt/issues/1548

### CODE REVIEW

File: examples/kubeval/imperative/.expected/config.yaml
Comment (frankfarzan) [inline examples/kubeval/imperative/.expected/config.yaml:7]: With structured results design, there's a change in plan for testing. We should have a few e2e tests in the kpt repo to test the presentation layer. But otherwise, we should only test the structured output which should provide all the information we want to test the functions robustly (depend on APIs not presentation layer). In particular, packages in the catalog repo shouldn't test the presentation layer. Otherwise, you'd get brittle change detector tests:

https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html

Changing the presentation layer in the future should involve updating just a few tests.

File: examples/kubeval/imperative/README.md
Comment (mikebz) [inline examples/kubeval/imperative/README.md:15]: very nice to show people how to do it!

Comment (mikebz) [reply examples/kubeval/imperative/README.md:15]: I would also maybe show multiple parameters if possible.  The next question people generally have is "do I use commas or spaces?"

File: examples/kubeval/imperative/.expected/config.yaml
Comment (mengqiy) [reply examples/kubeval/imperative/.expected/config.yaml:7]: @frankfarzan I remember we have that discussion in another PR. Are you suggesting hold this PR until that change lands?
@Shell32-Natsu What is the status? Any ETA?

File: examples/kubeval/imperative/README.md
Comment (mengqiy) [reply examples/kubeval/imperative/README.md:15]: I will update with multiple parameters.
Thanks for the suggestion.

File: examples/kubeval/imperative/.expected/config.yaml
Comment (mengqiy) [reply examples/kubeval/imperative/.expected/config.yaml]: Updated the PR to only test structured results

**Comment by mengqiy** [comment]
Tests related to kpt fn eval are failing because we have pick https://github.com/GoogleContainerTools/kpt/pull/1985.
I will update the eval tests

**Comment by mengqiy** [comment]
Rebased to pick up some new fn changes.

**Comment by mengqiy** [comment]
PTAL

File: examples/ensure-name-substring/advanced/README.md
Comment (phanimarupaka) [inline examples/ensure-name-substring/advanced/README.md:61]: Cat is deprecated. https://kpt.dev/reference/pkg/

File: examples/set-annotations/advanced/README.md
Comment (phanimarupaka) [inline examples/set-annotations/advanced/README.md:28]: Is this needed in kpt v1 ?

File: examples/ensure-name-substring/advanced/README.md
Comment (mengqiy) [reply examples/ensure-name-substring/advanced/README.md:61]: Is there a replacement?
Any suggestion?

File: examples/set-annotations/advanced/README.md
Comment (mengqiy) [reply examples/set-annotations/advanced/README.md:28]: In kpt v1, kpt live still understand it.
I want to ensure the example package is compatible with `kpt live` (IOW the package should be able to kpt live apply after the transformation).
Otherwise, the fn config CRD will be submitted to the server and cause an error.

File: examples/ensure-name-substring/advanced/README.md
Comment (phanimarupaka) [reply examples/ensure-name-substring/advanced/README.md:61]: linux cat specific files ?

File: examples/set-annotations/advanced/README.md
Comment (phanimarupaka) [reply examples/set-annotations/advanced/README.md:28]: Shouldn't kpt live also behave the same way as fn and identify such configs pointed from Kptfile ?

@frankfarzan

Comment (mengqiy) [reply examples/set-annotations/advanced/README.md:28]: +1

File: .github/workflows/ci.yaml
Comment (frankfarzan) [inline .github/workflows/ci.yaml:92]: is there an issue for this?

File: examples/apply-setters/imperative/README.md
Comment (frankfarzan) [inline examples/apply-setters/imperative/README.md:5]: why are we adding new examples in this PR? Especially given they're incomplete.

File: examples/enforce-gatekeeper/imperative/README.md
Comment (frankfarzan) [inline examples/enforce-gatekeeper/imperative/README.md:38]: You may not want to show the non-structured output, it's likely going to change, and these dos will become stale.

File: examples/enforce-gatekeeper/imperative/README.md
Comment (frankfarzan) [inline examples/enforce-gatekeeper/imperative/README.md:24]: I would just specify `--results-dir=tmp` and you skip this step

File: examples/enforce-gatekeeper/warning-only/README.md
Comment (frankfarzan) [inline examples/enforce-gatekeeper/warning-only/README.md:5]: What's "invalid configmap example"? Either link or drop the reference.

Comment (frankfarzan) [inline examples/enforce-gatekeeper/warning-only/README.md:5]: Please audit all the docs for go/kpt-docs-style-guide

e.g. style issue: configmap -> `ConfigMap`

File: examples/enforce-gatekeeper/warning-only/README.md
Comment (frankfarzan) [inline examples/enforce-gatekeeper/warning-only/README.md:58]: @droot why?

File: examples/set-annotations/imperative/.expected/diff.patch
Comment (frankfarzan) [inline examples/set-annotations/imperative/.expected/diff.patch:1]: General note:

We should have better file names than `resources`. Similar to naming this `configs`, it convey no information as everything is a resource, so it should be avoided.

File: examples/set-annotations/imperative/README.md
Comment (frankfarzan) [inline examples/set-annotations/imperative/README.md:3]: nit: These should be h2 (##).

File: examples/set-annotations/advanced/README.md
Comment (frankfarzan) [reply examples/set-annotations/advanced/README.md:28]: We shouldn't need `local-config` as live should understand the semantics. There's a tracking issue for this and @mortent is driving it in m3.

File: examples/set-annotations/imperative/README.md
Comment (mengqiy) [reply examples/set-annotations/imperative/README.md:3]: Does this comment conflict with the 2nd item in https://github.com/GoogleContainerTools/kpt/issues/1825?

File: examples/enforce-gatekeeper/warning-only/README.md
Comment (mengqiy) [reply examples/enforce-gatekeeper/warning-only/README.md:58]: Last update was 5+ days ago. The behavior may have changed already.
I will verify if this is still the latest behavior.

File: .github/workflows/ci.yaml
Comment (mengqiy) [reply .github/workflows/ci.yaml:92]: Given most behaviors of render and eval have become much more stable recently, I can change to use `next`:
`go get github.com/GoogleContainerTools/kpt@next`.

File: examples/apply-setters/imperative/README.md
Comment (mengqiy) [reply examples/apply-setters/imperative/README.md:5]: I can split the imperative examples out to a separate PR.

File: examples/kubeval/imperative/.expected/config.yaml
Comment (mengqiy) [reply examples/kubeval/imperative/.expected/config.yaml]: >@droot is adding support GoogleContainerTools/kpt#1923

> Can you coordinate with him?

Looking at the time stamp of this comment, I guess you intended to publish last week but forgot.

File: examples/set-annotations/imperative/.expected/diff.patch
Comment (mengqiy) [reply examples/set-annotations/imperative/.expected/diff.patch]: I'm not sure if `krm-resources.yaml` would be a better name.
Any suggested names?

**Comment by mengqiy** [comment]
e2e-ci is failing due to https://github.com/GoogleContainerTools/kpt/issues/2109

**Comment by mengqiy** [comment]
The last 2 commits are new changes since last review.

Commit `update expected stderr` is caused by GoogleContainerTools/kpt#2109. I will fix it in upstream kyaml. 
The CheckSite CI issue is tracked in https://github.com/GoogleContainerTools/kpt/issues/2112 and Eyob is going to tackle it this afternoon.

File: examples/ensure-name-substring/advanced/Kptfile
Comment (frankfarzan) [inline examples/ensure-name-substring/advanced/Kptfile:7]: will unstable be removed in the final docs?

File: examples/kubeval/simple/README.md
Comment (frankfarzan) [inline examples/kubeval/simple/README.md:40]: I would recommend to showing the the non-structured output which is likely to change (no backwards compatibility).

File: examples/kubeval/simple/README.md
Comment (frankfarzan) [inline examples/kubeval/simple/README.md:39]: Style guide says uses `shell` block quote everywhere. https://github.com/GoogleContainerTools/kpt/blob/next/docs/style-guides/docs.md

File: examples/set-annotations/simple/README.md
Comment (frankfarzan) [inline examples/set-annotations/simple/README.md:23]: https://github.com/GoogleContainerTools/kpt/blob/next/docs/style-guides/docs.md Rule 6

Please audit all the docs for style guides

File: tests/e2e.sh
Comment (frankfarzan) [inline tests/e2e.sh:20]: what's going on with this file?

File: tests/set-namespace/empty-input/simple/.expected/results.yaml
Comment (frankfarzan) [inline tests/set-namespace/empty-input/simple/.expected/results.yaml:8]: what's going on here?

Comment (mengqiy) [reply tests/set-namespace/empty-input/simple/.expected/results.yaml:8]: As mentioned in my earlier comment above.
https://github.com/GoogleContainerTools/kpt/issues/2109#issuecomment-849010352 is causing this issue. I will create a PR for it in kyaml this week. It will land and be release after some time. We will bump the kyaml dependencies in functions after that. I don't think we want to block this PR because of this minor issue.

File: examples/ensure-name-substring/advanced/Kptfile
Comment (mengqiy) [reply examples/ensure-name-substring/advanced/Kptfile:7]: `unstable` tag will be switched to the corresponding version tag in their release branches.
We will keep using unstable in the master branch.

**Review by frankfarzan — APPROVED** [review]
remaining comments are not blocking, feel free to follow up in this PR or separate PR.

**Comment by mengqiy** [comment]
Rebased to pick a fix for `CheckSite` CI.
The last commit contains the changes since last review.

File: tests/e2e.sh
Comment (mengqiy) [reply tests/e2e.sh]: We temporarily disable a test for contrib functions, since the test depends on the old kpt.
I will find some time to fix it in a followup PR.

**Pull request merged**
